### PR TITLE
Olimex stm32 e407 bugfixes

### PR DESCRIPTION
**Pull request by lowlander**
Several bug fixes for the olimex_stm32_e407 board. This should fix the compile bug you reported in Jira (ZEP-2452) by Anas. Trying to fix that bug I found several other issues, most of them no big deals. 

Except the "ospeed" one, without this Ethernet doesn't work anymore after Erwan's pinmux rework. With setting the ospeed it works again. Very interesting was that I have another board where Ethernet worked, the difference was that the working board used GPIOB for the ETH_TX pins and the olimex board uses GPIOG.  

@erwango  can you please look at the ospeed fix, since this might not just affect the Ethernet controller pins
@nashif  Anas could you check if this fixes your Jira bug

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
In the 'write gpio ospeed register' commit, can you update the commit message to explain why we need to do this.

**Comment by galak** [comment]
IN commit 'board: olimex_stm32_e407: fix default 48MHz clock divider' can you explain why in the commit message.

**Comment by lowlander** [comment]
@galak yeah NP. To the ospeed commit, I was hoping @erwango could comment on that because he actually already has a "ospeed" variable in the function, it just seems he forgot to write that to the chip register.  Now all output pin speeds are always at the lowest speed (00) which seem to cause problems with the ethernet pins, but apparently not on all boards (maybe an EMC issue)

So @erwango I am correct that setting the ospeed was simply forgotten, or was it intentionally left out?

**Comment by ydamigos** [comment]
I took a look in the PR #888 which changed the soc_gpio.c for the stm32 families.
Previously it contained the following code for the ospeed:
```
static u32_t __func_to_ospeed(int func)
{
	switch (func) {
	case STM32F4X_PIN_CONFIG_DRIVE_PUSH_PULL:
	case STM32F4X_PIN_CONFIG_DRIVE_PUSH_UP:
	case STM32F4X_PIN_CONFIG_DRIVE_PUSH_DOWN:
	case STM32F4X_PIN_CONFIG_DRIVE_OPEN_DRAIN:
	case STM32F4X_PIN_CONFIG_DRIVE_OPEN_UP:
	case STM32F4X_PIN_CONFIG_DRIVE_OPEN_DOWN:
	case STM32F4X_PIN_CONFIG_AF_PUSH_PULL:
	case STM32F4X_PIN_CONFIG_AF_PUSH_UP:
	case STM32F4X_PIN_CONFIG_AF_PUSH_DOWN:
	case STM32F4X_PIN_CONFIG_AF_OPEN_DRAIN:
	case STM32F4X_PIN_CONFIG_AF_OPEN_UP:
	case STM32F4X_PIN_CONFIG_AF_OPEN_DOWN:
		/* Force fast speed by default */
		return 0x2;
	}
	return 0;
}
```
And in the function stm32_gpio_configure
```
u32_t ospeed = __func_to_ospeed(conf);
```

```
	if (ospeed) {
		tmpreg = gpio->ospeed;
		tmpreg &= ~(0x3 << (pin * 2));
		tmpreg |= (ospeed << (pin * 2));
		gpio->ospeed = tmpreg;
	}
```

So as @lowlander mentions the question is if it was forgotten or not.

**Comment by lowlander** [comment]
@galak happy with the commit messges ?

**Comment by lowlander** [comment]
@ydamigos if @erwango  is back and can confirm it is just an oversight I can also add it to the other STM32 cpu's.

**Comment by galak** [comment]
@lowlander lets add ospeed setting to the other SoCs.  I think this is just on oversight and since we don't have anything that sets any STM32_OSPEEDR_* values, its not going to hurt anything.

File: arch/arm/soc/st_stm32/stm32f4/soc_gpio.c [hunk 1]
```diff
@@ -69,6 +69,9 @@ int stm32_gpio_configure(u32_t *base_addr, int pin, int conf, int altf)
 	scratch = gpio->mode & ~(STM32_MODER_MASK << pin_shift);
 	gpio->mode = scratch | (mode << pin_shift);
 
+	scratch = gpio->ospeed & ~(STM32_OSPEEDR_MASK << pin_shift);
```
Comment (galak) [inline arch/arm/soc/st_stm32/stm32f4/soc_gpio.c]: Lets fix this for f3 & l4 as well in this patch.

**Comment by galak** [comment]
@lowlander commits msg look good, if we fixup ospeed on F3/L4 will be good to merge.

**Comment by lowlander** [comment]
@galak since I don't have any F3 or L4 hardware I only did a "it compiles" test. If you have F3 or L4 hardware it would give me a better feeling if you could give it a try on actual hardware.

**Comment by ydamigos** [comment]
@lowlander I have a stm32f3_disco. What sample should I try?

**Comment by lowlander** [comment]
@ydamigos anything that uses GPIO (blinky?)

**Comment by ydamigos** [comment]
Tested samples/basic/{blinky, button} on stm32f3_disco board and they are working fine.

**Comment by lowlander** [comment]
@ydamigos thanx for testing, think L4 should be fine too then.

**Review by galak — APPROVED** [review]

**Pull request merged**